libdmtx/man/libdmtx.3
.\" $Id: libdmtx.3 771 2009-03-03 01:51:15Z mblaughton $
.\"
.\" Man page for the libdmtx project.
.\"
.\" To view: $ groff -man -T ascii libdmtx.3 | less
.\" To text: $ groff -man -T ascii libdmtx.3 | col -b | expand
.\"
.TH LIBDMTX 3 "March 2, 2009"
.SH NAME
libdmtx \- Data Matrix Encoding & Decoding Library 0.7.0
.SH SYNOPSIS
\fB#include <dmtx.h>\fP

cc file.c -ldmtx

.SH DESCRIPTION
\fIlibdmtx\fP is a software library that enables programs to read and write Data Matrix barcodes of the modern ECC200 variety. The library runs natively on several platforms, and can be accessed by multiple languages using the libdmtx language wrappers. The utility programs \fIdmtxread\fP and \fIdmtxwrite\fP provide a command line interface for libdmtx, and serve as a good reference for developers writing their own libdmtx-enabled programs.

Data Matrix barcodes store data as a pattern of ON and OFF modules (often black on white) in a grid pattern that resembles a checkerboard. Like other 2D symbologies, Data Matrix barcodes have a large data capacity compared to their traditional 1D cousins, and employ sophisticated error correction techniques. Data Matrix barcodes can be square or rectangle in shape, and offer several encodation schemes for optimized storage of text and/or binary data. The Data Matrix symbology was invented and released into the public domain by RVSI Acuity CiMatrix.

.SH ENCODING - Generating Data Matrix Barcodes
C/C++ programs can generate a barcode with just a few basic calls to libdmtx:

1. Call \fBdmtxEncodeCreate()\fP

Creates a new \fBDmtxEncode\fP structure and initializes the encoding process. This function must be called before using the other encoding functions.

2. Call \fBdmtxEncodeSetProp()\fP [optional]

Allows you to control specific aspects of the encoding behavior. If this function is not called, libdmtx will use the defaults set by \fBdmtxEncodeCreate()\fP above. The complementary function, \fBdmtxEncodeGetProp()\fP, allows you to detect the current settings.

3. Call either \fBdmtxEncodeDataMatrix()\fP or \fBdmtxEncodeDataMosaic()\fP

Call one of these functions to generate an image of the desired barcode type. Your program is responsible for dispatching the resulting output to its destination, whether that means displaying it on a screen, writing an image file, copying it elsewhere, etc...

4. Call \fBdmtxEncodeDestroy()\fP

Releases memory allocated during the encoding process.

.SH DECODING - Reading Data Matrix Barcodes
Barcode reading takes more steps than barcode generation, mainly because libdmtx must find a barcode region before it can decode the message. However, this too is a relatively simple process that uses 4 main structures:

\fBDmtxImage\fP holds image properties and a pointer to pixel data held by the calling program.

\fBDmtxDecode\fP holds values for controlling decode behavior and tracking scan progress. When scanning a new image, calling programs should always create a new \fBDmtxDecode\fP structure instead of reusing an old one.

\fBDmtxRegion\fP defines a 4-sided region in pixel coordinates. Regions may be found in almost any orientation, and their corners won't necessarily form right angles. libdmtx uses this structure to store the location of potential barcodes, which are then returned to the calling program one-at-a-time.

\fBDmtxMessage\fP holds the decoded message after being extracted from the barcode region. A successfully decoded region will produce exactly one message.

Use the following functions to find and decode Data Matrix barcodes:

1. Call \fBdmtxImageCreate()\fP

Creates and initializes a new \fBDmtxImage\fP structure using pixel data provided by the calling application. Parameters include a pointer to the existing pixel array, image width, height, and the pixel packing format.

2. Call \fBdmtxImageSetProp()\fP [optional]

Sets image properties to control the pixel mapping logic. These settings allow libdmtx to understand many native in-memory image layouts, thus preventing the extra work of transforming and copying data to a one-size-fits-all format. A \fBdmtxDecodeGetProp()\fP function is also available for detecting the current image properties.

3. Call \fBdmtxDecodeCreate()\fP

Creates and initializes a new \fBDmtxDecode\fP struct, which designates the image to be scanned and initializes the scan grid pattern. This function must be called before any other scanning functions.

4. Call \fBdmtxDecodeSetProp()\fP [optional]

Sets internal properties to control decoding behavior. This feature allows you to optimize performance and accuracy for specific image conditions. A \fBdmtxDecodeGetProp()\fP function is also available.

5. Call \fBdmtxRegionFindNext()\fP

Searches every pixel location in a grid pattern looking for potential barcode regions. A \fBDmtxRegion\fP is returned whenever a potential barcode region is found, or if the final pixel location has been scanned. Subsequent calls to this function will resume the search where the previous call left off.

6. Call either \fBdmtxDecodeMatrixRegion()\fP or \fBdmtxDecodeMosaicRegion()\fP

Extracts raw data from the barcode region and decodes the underlying message.

7. Call \fBdmtxMessageDestroy()\fP

Releases memory held by a \fBDmtxMessage\fP struct. The complementary function, \fBdmtxMessageCreate()\fP, is automatically called by \fBdmtxDecodeMatrixRegion()\fP and therefore is not normally used by the calling program.

8. Call \fBdmtxRegionDestroy()\fP

Releases memory held by a \fBDmtxRegion\fP struct. The complementary function, \fBdmtxRegionCreate()\fP, is automatically called by \fBdmtxRegionFindNext()\fP (actually \fBdmtxRegionScanPixel()\fP) and therefore is not normally used by the calling program.

9. Call \fBdmtxDecodeDestroy()\fP

Releases memory held by a \fBDmtxDecode\fP struct. This is the complementary function to \fBdmtxDecodeCreate()\fP.

10. Call \fBdmtxImageDestroy()\fP

Releases memory held by a \fBDmtxImage\fP struct, excluding the pixel array passed to \fBdmtxImageCreate()\fP. The calling program is responsible for releasing the pixel array memory, if required.

.SH EXAMPLE PROGRAM

This example program (available as simple_test.c in the source package) demonstrates \fIlibdmtx\fP functionality in both directions: encoding and decoding. It creates a Data Matrix barcode in memory, reads it back, and prints the decoded message. The final output message should match the original input string.

  #include <stdlib.h>
  #include <stdio.h>
  #include <string.h>
  #include <assert.h>
  #include <dmtx.h>

  int
  main(int argc, char *argv[])
  {
     size_t          width, height, bytesPerPixel;
     unsigned char   str[] = "30Q324343430794<OQQ";
     unsigned char  *pxl;
     DmtxEncode     *enc;
     DmtxImage      *img;
     DmtxDecode     *dec;
     DmtxRegion     *reg;
     DmtxMessage    *msg;

     fprintf(stdout, "input:  \\"%s\\"\\n", str);

     /* 1) ENCODE a new Data Matrix barcode image (in memory only) */

     enc = dmtxEncodeCreate();
     assert(enc != NULL);
     dmtxEncodeDataMatrix(enc, strlen(str), str);

     /* 2) COPY the new image data before releasing encoding memory */

     width = dmtxImageGetProp(enc->image, DmtxPropWidth);
     height = dmtxImageGetProp(enc->image, DmtxPropHeight);
     bytesPerPixel = dmtxImageGetProp(enc->image, DmtxPropBytesPerPixel);

     pxl = (unsigned char *)malloc(width * height * bytesPerPixel);
     assert(pxl != NULL);
     memcpy(pxl, enc->image->pxl, width * height * bytesPerPixel);

     dmtxEncodeDestroy(&enc);

     /* 3) DECODE the Data Matrix barcode from the copied image */

     img = dmtxImageCreate(pxl, width, height, DmtxPack24bppRGB);
     assert(img != NULL);

     dec = dmtxDecodeCreate(img, 1);
     assert(dec != NULL);

     reg = dmtxRegionFindNext(dec, NULL);
     if(reg != NULL) {
        msg = dmtxDecodeMatrixRegion(dec, reg, DmtxUndefined);
        if(msg != NULL) {
           fputs("output: \\"", stdout);
           fwrite(msg->output, sizeof(unsigned char), msg->outputIdx, stdout);
           fputs("\\"\\n", stdout);
           dmtxMessageDestroy(&msg);
        }
        dmtxRegionDestroy(&reg);
     }

     dmtxDecodeDestroy(&dec);
     dmtxImageDestroy(&img);
     free(pxl);

     exit(0);
  }

.SH "SEE ALSO"
\fIdmtxread\fP(1), \fIdmtxwrite\fP(1), \fIdmtxquery\fP(1)
.SH STANDARDS
ISO/IEC 16022:2000
.PP
ANSI/AIM BC11 ISS
.SH BUGS
Email bug reports to mike@dragonflylogic.com
.SH AUTHOR
Copyright (C) 2008, 2009 Mike Laughton
.\" end of man page


libdmtx/wrapper/net/LibDmtx.cs
/*
libdmtx-net - .NET wrapper for libdmtx

Copyright (C) 2009 Joseph Ferner / Tom Vali

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

Contact: libdmtx@fernsroth.com
*/

/* $Id: LibDmtx.cs 768 2009-03-02 01:42:47Z joegtp $ */

using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Drawing;
using System.Drawing.Imaging;
using System.Text;

namespace Libdmtx {
    /// <summary>
    /// Wrapper for decoding and encoding DataMatrix barcodes.
    /// </summary>
    public static class Dmtx {
        const byte RETURN_NO_MEMORY = 1;
        const byte RETURN_INVALID_ARGUMENT = 2;
        const byte RETURN_ENCODE_ERROR = 3;
        public const int DmtxUndefined = -1; // defined in "dmtx.h"

        /// <summary>
        /// Gets the version of the underlying libdmtx used.
        /// </summary>
        public static string Version {
            get {
                return DmtxVersion();
            }
        }

        /// <summary>
        /// Decodes a bitmap returning all symbols found in the image.
        /// </summary>
        /// <param name="b">The bitmap to decode.</param>
        /// <param name="options">The options used for decoding.</param>
        /// <returns>An array of decoded symbols, one for each symbol found.</returns>
        /// <example>
        /// This example shows a basic decoding.
        /// <code>
        ///   Bitmap bm = (Bitmap)Bitmap.FromFile("barcode.bmp");
        ///   DecodeOptions decodeOptions = new DecodeOptions();
        ///   DmtxDecoded[] decodeResults = Dmtx.Decode(bm, decodeOptions);
        ///   for (int i = 0; i &lt; decodeResults.Length; i++) {
        ///     string str = Encoding.ASCII.GetString(decodeResults[i].Data).TrimEnd('\0');
        ///     Console.WriteLine("Decode " + i + ": \"" + str + "\"");
        ///   }
        /// </code>
        /// </example>
        public static DmtxDecoded[] Decode(Bitmap b, DecodeOptions options) {
            List<DmtxDecoded> results = new List<DmtxDecoded>();
            Decode(b, options, delegate(DmtxDecoded d) { results.Add(d); });
            return results.ToArray();
        }

        public static DmtxDecoded[] Decode(
            Bitmap b,
            DecodeOptions options,
            DiagnosticImageStyles diagnosticImageStyle, out Bitmap diagnosticImage) {
            List<DmtxDecoded> results = new List<DmtxDecoded>();
            Bitmap diagnosticImageTemp = null;
            Decode(
                b,
                options,
                delegate(DmtxDecoded d) { results.Add(d); },
                diagnosticImageStyle,
                delegate(Bitmap di) { diagnosticImageTemp = di; });
            diagnosticImage = diagnosticImageTemp;
            return results.ToArray();
        }

        public delegate void DecodeCallback(DmtxDecoded decoded);

        public delegate void DecodeDiagnosticImageCallback(Bitmap diagnosticImage);

        public static void Decode(Bitmap b, DecodeOptions options, DecodeCallback Callback) {
            Decode(b, options, Callback, 0, null);
        }

        public static void Decode(
            Bitmap b,
            DecodeOptions options,
            DecodeCallback Callback,
            DiagnosticImageStyles diagnosticImageStyle, DecodeDiagnosticImageCallback DiagnosticImageCallback) {
            Exception decodeException = null;
            byte status;
            try {
                int bitmapStride;
                byte[] pxl = BitmapToByteArray(b, out bitmapStride);

                DmtxDiagnosticImageCallback diagnosticImageCallbackParam = null;
                if (DiagnosticImageCallback != null) {
                    diagnosticImageCallbackParam = delegate(IntPtr data, uint totalBytes, uint headerBytes) {
                        try {
                            byte[] pnmData = new byte[totalBytes];
                            Marshal.Copy(data, pnmData, 0, pnmData.Length);
                            using (MemoryStream pnmInputStream = new MemoryStream(pnmData)) {
                                Bitmap bm = PnmToBitmap(pnmInputStream);
                                DiagnosticImageCallback(bm);
                            }
                        } catch (Exception ex) {
                            decodeException = ex;
                        }
                    };
                }

                status = DmtxDecode(
                    pxl,
                    (UInt32)b.Width,
                    (UInt32)b.Height,
                    (UInt32)bitmapStride,
                    options,
                    diagnosticImageCallbackParam, diagnosticImageStyle,
                    delegate(DecodedInternal dmtxDecodeResult) {
                        DmtxDecoded result;
                        try {
                            result = new DmtxDecoded();
                            result.Corners = dmtxDecodeResult.Corners;
                            result.SymbolInfo = dmtxDecodeResult.SymbolInfo;
                            result.Data = new byte[dmtxDecodeResult.DataSize];
                            for (int dataIdx = 0; dataIdx < dmtxDecodeResult.DataSize; dataIdx++) {
                                result.Data[dataIdx] = Marshal.ReadByte(dmtxDecodeResult.Data, dataIdx);
                            }
                            Callback(result);
                            return true;
                        } catch (Exception ex) {
                            decodeException = ex;
                            return false;
                        }
                    });
            } catch (Exception ex) {
                throw new DmtxException("Error calling native function.", ex);
            }
            if (decodeException != null) {
                throw decodeException;
            }
            if (status == RETURN_NO_MEMORY) {
                throw new DmtxOutOfMemoryException("Not enough memory.");
            } else if (status == RETURN_INVALID_ARGUMENT) {
                throw new DmtxInvalidArgumentException("Invalid options configuration.");
            } else if (status > 0) {
                throw new DmtxException("Unknown error.");
            }
        }

        private static byte[] BitmapToByteArray(Bitmap b, out int stride) {
            Rectangle rect = new Rectangle(0, 0, b.Width, b.Height);
            BitmapData bd = b.LockBits(rect, ImageLockMode.ReadOnly, PixelFormat.Format24bppRgb);
            try {
                byte[] pxl = new byte[bd.Stride * b.Height];
                Marshal.Copy(bd.Scan0, pxl, 0, bd.Stride * b.Height);
                stride = bd.Stride;
                return pxl;
            } finally {
                b.UnlockBits(bd);
            }
        }

        /// <summary>
        /// Encodes data into a DataMatrix barcode.
        /// </summary>
        /// <param name="data">The data to encode.</param>
        /// <param name="options">The options used for encoding.</param>
        /// <returns>The results from encoding.</returns>
        /// <example>
        /// This example shows a basic encoding.
        /// <code>
        ///   byte[] data = Encoding.ASCII.GetBytes("test");
        ///   EncodeOptions o = new EncodeOptions();
        ///   DmtxEncoded encodeResults = Dmtx.Encode(data, o);
        ///   encodeResults.Bitmap.Save("barcode.bmp", ImageFormat.Bmp);
        /// </code>
        /// </example>
        public static DmtxEncoded Encode(byte[] data, EncodeOptions options) {
            IntPtr result;
            byte status;
            try {
                status = DmtxEncode(data, (UInt16)data.Length, out result, options);
            } catch (Exception ex) {
                throw new DmtxException("Encoding error.", ex);
            }
            if (status == RETURN_NO_MEMORY) {
                throw new DmtxOutOfMemoryException("Not enough memory.");
            } else if (status == RETURN_INVALID_ARGUMENT) {
                throw new DmtxInvalidArgumentException("Invalid options configuration.");
            } else if (status == RETURN_ENCODE_ERROR) {
                throw new DmtxException("Error while encoding.");
            } else if ((status > 0) || (result == IntPtr.Zero)) {
                throw new DmtxException("Unknown error.");
            }

            DmtxEncoded ret;
            EncodedInternal intResult = null;
            try {
                intResult = (EncodedInternal)Marshal.PtrToStructure(result, typeof(EncodedInternal));
                ret = new DmtxEncoded();
                ret.SymbolInfo = intResult.SymbolInfo;
                ret.Bitmap = new Bitmap((int)intResult.Width, (int)intResult.Height, PixelFormat.Format24bppRgb);
                Rectangle rect = new Rectangle(0, 0, ret.Bitmap.Width, ret.Bitmap.Height);
                BitmapData bd = ret.Bitmap.LockBits(rect, ImageLockMode.WriteOnly, PixelFormat.Format24bppRgb);
                try {
                    DmtxCopyEncodeResult(intResult.Data, (UInt32)Math.Abs(bd.Stride), bd.Scan0);
                } finally {
                    ret.Bitmap.UnlockBits(bd);
                }
            } catch (Exception ex) {
                throw new DmtxException("Error parsing encode result.", ex);
            } finally {
                try {
                    if (intResult != null) {
                        DmtxFreeEncodeResult(intResult.Data);
                    }
                } catch (Exception ex) {
                    throw new DmtxException("Error freeing memory.", ex);
                }
            }
            return ret;
        }

        public static Bitmap PnmToBitmap(Stream pnmInputStream) {
            // read header
            if (ReadLine(pnmInputStream) != "P6") {
                throw new Exception("Invalid PNM file format. Header was invalid.");
            }
            string widthHeightLine = ReadLine(pnmInputStream);
            string[] widthHeightParts = widthHeightLine.Split(' ');
            int width = int.Parse(widthHeightParts[0]);
            int height = int.Parse(widthHeightParts[1]);
            ReadLine(pnmInputStream); // ignore max value

            Bitmap result = new Bitmap(width, height, PixelFormat.Format24bppRgb);

            // transfer pixel data to bitmap
            Rectangle rect = new Rectangle(0, 0, width, height);
            BitmapData bd = result.LockBits(rect, ImageLockMode.WriteOnly, PixelFormat.Format24bppRgb);
            try {
                byte[] buffer = new byte[height * bd.Stride];
                for (int y = 0; y < height; y++) {
                    int rowOffset = y * bd.Stride;
                    pnmInputStream.Read(buffer, rowOffset, width * 3);
                }

                Marshal.Copy(buffer, 0, bd.Scan0, buffer.Length);
            } finally {
                result.UnlockBits(bd);
            }

            return result;
        }

        private static string ReadLine(Stream stream) {
            StringBuilder result = new StringBuilder();
            int b;
            while ((b = stream.ReadByte()) != '\n') {
                result.Append((char)b);
            }
            return result.ToString();
        }

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        private delegate bool DmtxDecodeCallback(DecodedInternal dmtxDecodeResult);

        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
        private delegate void DmtxDiagnosticImageCallback(IntPtr data, uint totalBytes, uint headerSize);

        [DllImport("libdmtx.dll", EntryPoint = "dmtx_decode")]
        private static extern byte
        DmtxDecode(
            [In] byte[] image,
            [In] UInt32 width,
            [In] UInt32 height,
            [In] UInt32 bitmapStride,
            [In] DecodeOptions options,
            [In] DmtxDiagnosticImageCallback diagnosticImageCallback,
            [In] DiagnosticImageStyles diagnosticImageStyle,
            [In] DmtxDecodeCallback decodeCallback);

        [DllImport("libdmtx.dll", EntryPoint = "dmtx_encode")]
        private static extern byte
        DmtxEncode(
            [In] byte[] plain_text,
            [In] UInt16 text_size,
            [Out] out IntPtr result,
            [In] EncodeOptions options);

        [DllImport("libdmtx.dll", EntryPoint = "dmtx_copy_encode_result")]
        private static extern void
        DmtxCopyEncodeResult(
            [In] IntPtr data,
            [In] UInt32 stride,
            [In] IntPtr bitmap);

        [DllImport("libdmtx.dll", EntryPoint = "dmtx_free_encode_result")]
        private static extern void
        DmtxFreeEncodeResult([In] IntPtr data);

        [DllImport("libdmtx.dll", EntryPoint = "dmtx_version", CharSet = CharSet.Ansi)]
        private static extern string
        DmtxVersion();
    }

    public enum DiagnosticImageStyles : uint {
        Default = 0
    }

    /// <summary>
    /// Enumeration of symbol sizes.
    /// </summary>
    public enum CodeSize : short {
        SymbolRectAuto = -3,
        SymbolSquareAuto = -2,
        SymbolShapeAuto = -1,
        Symbol10x10 = 0,
        Symbol12x12 = 1,
        Symbol14x14,
        Symbol16x16,
        Symbol18x18,
        Symbol20x20,
        Symbol22x22,
        Symbol24x24,
        Symbol26x26,
        Symbol32x32,
        Symbol36x36,
        Symbol40x40,
        Symbol44x44,
        Symbol48x48,
        Symbol52x52,
        Symbol64x64,
        Symbol72x72,
        Symbol80x80,
        Symbol88x88,
        Symbol96x96,
        Symbol104x104,
        Symbol120x120,
        Symbol132x132,
        Symbol144x144,
        Symbol8x18,
        Symbol8x32,
        Symbol12x26,
        Symbol12x36,
        Symbol16x36,
        Symbol16x48
    }

    /// <summary>
    /// Symbol type.
    /// </summary>
    public enum CodeType : ushort {
        DataMatrix = 0,

        /// <summary>
        /// Composite of multiple data matrixes.
        /// </summary>
        Mosaic = 1
    }

    /// <summary>
    /// Encodings applied to the data.
    /// </summary>
    public enum Scheme : ushort {
        /// <summary>
        /// ASCII character 0 to 127. 1 byte per CW.
        /// </summary>
        Ascii = 0,

        /// <summary>
        /// Upper-case alphanumeric. 1.5 byte per CW.
        /// </summary>
        C40 = 1,

        /// <summary>
        /// Lower-case alphanumeric. 1.5 byte per CW.
        /// </summary>
        Text,

        /// <summary>
        /// ANSI X12. 1.5 byte per CW.
        /// </summary>
        X12,

        /// <summary>
        /// ASCII character 32 to 94. 1.33 bytes per CW.
        /// </summary>
        Edifact,

        /// <summary>
        /// ASCII character 0 to 255. 1 byte per CW.
        /// </summary>
        Base256,
        AutoBest,
        AutoFast
    }

    /// <summary>
    /// Options used for decoding using <see cref="Dmtx.Decode(Bitmap,DecodeOptions)"/>
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public class DecodeOptions {
        public Int16 EdgeMin = Dmtx.DmtxUndefined;
        public Int16 EdgeMax = Dmtx.DmtxUndefined;
        public Int16 ScanGap = Dmtx.DmtxUndefined;
        public Int16 SquareDevn = Dmtx.DmtxUndefined;
        public Int32 TimeoutMS = Dmtx.DmtxUndefined;
        public CodeSize SizeIdxExpected = CodeSize.SymbolShapeAuto;
        public Int16 EdgeThresh = Dmtx.DmtxUndefined;
        public Int16 MaxCodes = Dmtx.DmtxUndefined;
        public Int16 XMin = Dmtx.DmtxUndefined;
        public Int16 XMax = Dmtx.DmtxUndefined;
        public Int16 YMin = Dmtx.DmtxUndefined;
        public Int16 YMax = Dmtx.DmtxUndefined;
        public Int16 CorrectionsMax = Dmtx.DmtxUndefined;
        public CodeType CodeType = CodeType.DataMatrix;
        public Int16 Shrink = 1;
    }

    /// <summary>
    /// Options used for encoding using <see cref="Dmtx.Encode"/>.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public class EncodeOptions {
        /// <summary>
        /// Whitespace padding in pixels around symbol.
        /// </summary>
        public UInt16 MarginSize;

        /// <summary>
        /// The size in pixels of each element in the symbol.
        /// </summary>
        public UInt16 ModuleSize;

        /// <summary>
        /// The encoding to apply to the symbol.
        /// </summary>
        public Scheme Scheme;
        public UInt16 Rotate;

        /// <summary>
        /// Size of the symbol to generate.
        /// </summary>
        public CodeSize SizeIdx;

        /// <summary>
        /// Type of symbol to generate.
        /// </summary>
        public CodeType CodeType;

        public EncodeOptions() {
            MarginSize = 10;
            ModuleSize = 5;
            Scheme = Scheme.Ascii;
            Rotate = 0;
            SizeIdx = CodeSize.SymbolSquareAuto;
            CodeType = CodeType.DataMatrix;
        }
    }

    /// <summary>
    /// 2D coordinate.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public class DmtxPoint {
        public UInt16 X;
        public UInt16 Y;
    }

    /// <summary>
    /// The four corners in 2D space of the symbol.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public class Corners {
        public DmtxPoint Corner0;
        public DmtxPoint Corner1;
        public DmtxPoint Corner2;
        public DmtxPoint Corner3;
    }

    /// <summary>
    /// Information about the DMTX symbol.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public class SymbolInfo {
        public UInt16 Rows;
        public UInt16 Cols;
        public UInt16 Capacity;
        public UInt16 DataWords;
        public UInt16 PadWords;
        public UInt16 ErrorWords;
        public UInt16 HorizDataRegions;
        public UInt16 VertDataRegions;
        public UInt16 InterleavedBlocks;
        public UInt16 Angle;
    }

    /// <summary>
    /// Returned from <see cref="Dmtx.Decode(Bitmap,DecodeOptions)"/>.
    /// </summary>
    public class DmtxDecoded {
        /// <summary>
        /// Information about the symbol that was decoded.
        /// </summary>
        public SymbolInfo SymbolInfo;
        public Corners Corners;

        /// <summary>
        /// The data contained in the symbol. If the contains a string
        /// use the following code to convert it.
        /// <code>
        /// string str = Encoding.ASCII.GetString(decodeResults[0].Data).TrimEnd('\0');
        /// </code>
        /// </summary>
        public byte[] Data;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal class DecodedInternal {
        public SymbolInfo SymbolInfo;
        public Corners Corners;
        public IntPtr Data;
        public UInt32 DataSize;
    }

    /// <summary>
    /// Returned from <see cref="Dmtx.Encode"/>.
    /// </summary>
    public class DmtxEncoded {
        /// <summary>
        /// Information about the symbol that was created.
        /// </summary>
        public SymbolInfo SymbolInfo;

        /// <summary>
        /// The bitmap containing the symbol.
        /// </summary>
        public Bitmap Bitmap;
    }

    [StructLayout(LayoutKind.Sequential)]
    internal class EncodedInternal {
        public SymbolInfo SymbolInfo;
        public UInt32 Width;
        public UInt32 Height;
        public IntPtr Data;
    }

    /// <summary>
    /// Base Dmtx exception.
    /// </summary>
    public class DmtxException : ApplicationException {
        public DmtxException() { }
        public DmtxException(string message) : base(message) { }
        public DmtxException(string message, Exception innerException)
            : base(message, innerException) { }
    }

    /// <summary>
    /// Out of memory exception.
    /// </summary>
    public class DmtxOutOfMemoryException : DmtxException {
        public DmtxOutOfMemoryException() { }
        public DmtxOutOfMemoryException(string message) : base(message) { }
        public DmtxOutOfMemoryException(string message, Exception innerException)
            : base(message, innerException) { }
    }

    // Invalid argument exception.
    public class DmtxInvalidArgumentException : DmtxException {
        public DmtxInvalidArgumentException() { }
        public DmtxInvalidArgumentException(string message) : base(message) { }
        public DmtxInvalidArgumentException(string message, Exception innerException)
            : base(message, innerException) { }
    }
}